src/annotation.js

```diff
@@ -160,18 +160,18 @@ function draw(chart, caller, clip) {
   });
 }
 
-function drawElements(ctx, elements, caller) {
+function drawElements(ctx, elements, caller, parent) {
   for (const el of elements) {
     if (el.options.drawTime === caller) {
-      el.draw(ctx);
+      el.draw(ctx, parent);
     }
   }
 }
 
 function drawSubElements(ctx, elements, caller) {
   for (const el of elements) {
     if (isArray(el.elements)) {
-      drawElements(ctx, el.elements, caller);
+      drawElements(ctx, el.elements, caller, el);
```

Comment:
I'm not sure why I made it like this, where all of the sub elements are drawn after all of the annotations. If some of annotations overlap, all of the sub elements will be on top. This is another reason why I suggested the refactoring of the draw loop.

Reply:
> I'm not sure why I made it like this, where all of the sub elements are drawn after all of the annotations. If some of annotations overlap, all of the sub elements will be on top. This is another reason why I suggested the refactoring of the draw loop.

This is not 100% correct. Every sub-element has got own `drawTime` and therefore will be drawn based on that, like in the past. For instance, if the label is configured with `beforeDraw` and the element `afterDraw`, the element is drawn after the label.

Reply:
See line `if (el.options.drawTime === caller) {`.

```javascript
function drawElements(ctx, elements, caller) {
  for (const el of elements) {
    if (el.options.drawTime === caller) {
      el.draw(ctx);
    }
  }
}
```

Reply:
Right. My analysis only applies if all have the same drawTime. But the behavior could be unexpected anyway. But this can be changed separately if needed.

Reply:
> My analysis only applies if all have the same drawTime

See issue #175. In my opionion we could add a `z` option where, when the `drawTime` are equals, you can define the z-index.
But, afaik, you would like to add this logic to all chart elements instead to apply only to this plugin.
Anyway we could get all visible elements and all their subelements, extracting by drawtime, and then draw them (1 cycle).

Reply:
I think it can be done just by combining the two loops.

```js
for (const el of elements) {
  if (el.options.drawTime === caller) { 
    el.draw(ctx, chartArea); 
  }
  if (el.elements && el.elements.length) {
    const box = 'getBoundingBox' in el ? el.getBoundingBox() : chartArea;
    for (const sub of el.elements || []) {
      if (sub.options.drawTime === caller) { 
        sub.draw(ctx, box); 
      }
    }
  }
}

Reply:
To the z-index, we could add it here and adopt chart.js z-index when/if its available, but that should again be a separate pr.

Reply:
> I think it can be done just by combining the two loops.

I'm testing it, I have done it similar to your proposal.

Reply:
> To the z-index, we could add it here and adopt chart.js z-index when/if its available, but that should again be a separate pr.

OK!

Reply:
> I'm not sure why I made it like this, where all of the sub elements are drawn after all of the annotations.

I think you had done it to use recursively the drawElement function, that makes sense anyway.

Reply:
I don't think we should use chartArea by default. The annotations could be set with `clip: true` and therefore the area is not correct, I guess. Maybe canvas area.

Reply:
> I'm not sure why I made it like this, where all of the sub elements are drawn after all of the annotations. If some of annotations overlap, all of the sub elements will be on top. This is another reason why I suggested the refactoring of the draw loop.

@kurkle coding the line label as element, maybe I have understood why you implemented in that way. With new cycle to draw elements and sub elements, the label of the one are drawn after every own line element, This have got a side effect that when the `drawTime` is the same, the labels are NOT on top anymore but they could be crosses by another annotation.
Before this update, the labels were always drawn after all elements and therefore the labels were never overlapped by any annotation. 
Now, I discovered, during the tests (https://github.com/chartjs/chartjs-plugin-annotation/blob/master/test/fixtures/line/label_drawTime.js), I see this new behavior.